### PR TITLE
Add: "Software Tools: Rewriting History in Git"

### PR DESCRIPTION
**Pull request by JasonW029**
Adds a page detailing how to rewrite history in Git, which is a more advanced topic that helps keep your history clean.

Covers amending commits; soft, mixed, and hard resetting; and force-pushing changes to the remote.

Addresses: https://github.com/learning-software-engineering/learning-software-engineering.github.io/issues/404

### CODE REVIEW

**Review by SirRAL — COMMENTED** [review]
Very well written, with great and clear explanations! I enjoyed the intro very much; I think it is a great pitch for learning how to change history in Git. I mostly pointed out some parts that could be slightly expanded upon or some things that could be better emphasized, but overall, a very good job done on a complicated topic.

File: Topics/Software_Tools/Git/Git_Rewriting_History.md [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+And your history just looks prettier. You may not even need a better reason than that.
+
+## How Do You Rewrite History?
+
+### A Word of Warning: Commits are Overwritten
+
+Whenever you rewrite history in Git, you actually **overwrite** commits you have already made. This means that even if you think you're modifying a commit, you're actually replacing that commit with a new one, and deleting the old one.
+
+This means you get a new commit hash, and the new commit will be unrecognizable from any person who pulled the old commit. This causes Git to think that your histories have diverged, which will huge issues when merging. So, unless you know what you're doing, you should only rewrite history **before** you pushed your changes to the remote repository.
```
[the remark this answers is not in the document]
Comment (JasonW029) [reply Topics/Software_Tools/Git/Git_Rewriting_History.md]: I changed the line from "unless you know what you're doing" to a more descriptive "unless you're the only one who will touch your code".

File: Topics/Software_Tools/Git/Git_Rewriting_History.md [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+If you want to replace the old commit message, use:
+```
+ git commit --amend -m “<commit message here>"
+```
+
+Remember, this only affects the very last commit you made. If you want to fixup an older commit, you'll need the power of `git rebase`!
+
+### Discarding the Last Commits: `git reset`
+
+Git’s reset mechanic has a couple of modes that provide different use-cases: if you want to drop, squash, or remove (but keep the changes) from the last couple of commits, then `git reset` is the command for you.
```
[the remark this answers is not in the document]
Comment (JasonW029) [reply Topics/Software_Tools/Git/Git_Rewriting_History.md]: Added a new paragraph for this.

File: Topics/Software_Tools/Git/Git_Rewriting_History.md [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+It will reset your HEAD to the given commit AND wipe your staging area, but your local directory is untouched. In effect, this means that it keeps your local changes and the changes from the removed commits, but they are unstaged rather than automatically staged like in the “soft” reset.
+
+### For Dropping Commits: `git reset --hard`
+
+This is the most dangerous one of all. This resets you to the state of the given commit, throwing away your local changes and the changes from the removed commits, so it’s great if you want to start off on a clean slate exactly at the state of the given commit.
+
+It will reset your HEAD to the given commit, wipe your staging area, and reset your local directory. This is a reset in every sense of the word, and your changes are actually gone! Be very careful.
+
+Remember, this is only useful if you want to reset to the state of a previous commit. If you want granular control over which commits to drop between now and some point in time, read on for the section on `git rebase`!
```
[the remark this answers is not in the document]
Comment (JasonW029) [reply Topics/Software_Tools/Git/Git_Rewriting_History.md]: Oops, sure is. Removed.

File: Topics/Software_Tools/Git/Git_Rewriting_History.md [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+Git’s reset mechanic has a couple of modes that provide different use-cases: if you want to drop, squash, or remove (but keep the changes) from the last couple of commits, then `git reset` is the command for you.
+
+We’ll go over 3 modes: soft, mixed, and hard. In short, soft keeps and stages the resetted changes, mixed keeps the resetted changes, and hard will remove the resetted changes.
+
+
+### For Squashing Commits: `git reset --soft`
+
+`git reset --soft` is great for you if you’re looking to remove commits, but still keep the changes that were present in those commits in your staging area. This is useful if you want to squash the last couple of commits you made into a single commit.
+
+It will reset your HEAD to the given commit, but your local changes and the changes from those commits will remain and they’ll be staged. This means that using `git commit` right after soft-resetting will immediately re-commit the changes you just reset, since those resetted changes are still in your staging area.
```
[the remark this answers is not in the document]
Comment (JasonW029) [reply Topics/Software_Tools/Git/Git_Rewriting_History.md]: Added some wording for this.

File: Topics/Software_Tools/Git/Git_Rewriting_History.md [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+ git commit --amend -m “<commit message here>"
+```
+
+Remember, this only affects the very last commit you made. If you want to fixup an older commit, you'll need the power of `git rebase`!
+
+### Discarding the Last Commits: `git reset`
+
+Git’s reset mechanic has a couple of modes that provide different use-cases: if you want to drop, squash, or remove (but keep the changes) from the last couple of commits, then `git reset` is the command for you.
+
+We’ll go over 3 modes: soft, mixed, and hard. In short, soft keeps and stages the resetted changes, mixed keeps the resetted changes, and hard will remove the resetted changes.
```
[the remark this answers is not in the document]
Comment (JasonW029) [reply Topics/Software_Tools/Git/Git_Rewriting_History.md]: Added a command prototype for each of soft, mixed, and hard reset.

**Review by SirRAL — APPROVED** [review]
All requested changes have been made, and I'm very satisfied with what has been written. Good work!

**Pull request merged**